sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -200,6 +203,7 @@ class Column(val expr: Expression) extends Logging {
     // Wait until the struct is resolved. This will generate a nicer looking alias.
     case struct: CreateNamedStruct => UnresolvedAlias(struct)
 
+    case expr: Expression if existsUnresolvedExpr(expr) => UnresolvedAlias(expr)
     case expr: Expression => Alias(expr, toPrettySQL(expr))()
```

Comment:
Firstly, I tried to make all `Expression` use `UnresolvedAlias` but it failed with some test about `cast`. Before that `df.selectExpr("id1 == null").schema.head.name` is `(id1 = NULL)`, after that it changed to `(id1 = CAST(NULL AS INT))`.

So this code only change `UnresolvedExtractValue` to narrow it.
cc @maropu @cloud-fan

Reply:
Could you change the PR title into more specific one? I think the current one looks more general than it actually is.

Reply:
Changed it.

Reply:
Shall we strip auto-generated cast when resolving `UnresolvedAlias`?

Reply:
you mean add `case c @ Cast(e: Expression, _, _) => Alias(c, toPrettySQL(e))()` in `ResolveAliases` ?
```
private def assignAliases(exprs: Seq[NamedExpression]) = {
      exprs.map(_.transformUp { case u @ UnresolvedAlias(child, optGenAliasFunc) =>
          child match {
            case ne: NamedExpression => ne
            case go @ GeneratorOuter(g: Generator) if g.resolved => MultiAlias(go, Nil)
            case e if !e.resolved => u
            case g: Generator => MultiAlias(g, Nil)
            case c @ Cast(ne: NamedExpression, _, _) => Alias(c, ne.name)()
            case e: ExtractValue => Alias(e, toPrettySQL(e))()
            case e if optGenAliasFunc.isDefined =>
              Alias(child, optGenAliasFunc.get.apply(e))()
            case e => Alias(e, toPrettySQL(e))()
          }
        }
      ).asInstanceOf[Seq[NamedExpression]]
    }
```

I think that's ok logiclly but may affect many code place.

Reply:
We need a way to distinguish auto-generated cast and user-specified cast, probably via `TreeNodeTag`.

Reply:
Seems far from this PR, will do this in an another PR.

Reply:
Let's do that PR first. It looks to me that this PR fixes an issue but introduces another issue (the extra cast).

Reply:
Yeah I think so.